File: src/error.test.ts
import {
	HttpError,
	NotFoundError,
	ForbiddenError,
	ValidationError,
	UnauthorizedError,
	UnimplementedFunctionError,
} from './error';
import assert from 'assert';
import { describe, it } from 'node:test';

describe('HttpError Classes', { concurrency: true }, () => {
	describe('HttpError', () => {
		it('should set the statusCode and message correctly', () => {
			const error = new HttpError(500, 'Oh no, something went wrong!');
			assert.strictEqual(error.statusCode, 500);
			assert.strictEqual(error.message, 'Oh no, something went wrong!');
		});

		it('should set default statusCode and message', () => {
			const error = new HttpError();
			assert.strictEqual(error.statusCode, 500);
			assert.strictEqual(error.message, 'Oh no, something went wrong!');
		});
	});

	describe('ForbiddenError', { concurrency: true }, () => {
		it('should set the statusCode to 403 and message correctly', () => {
			const error = new ForbiddenError('Forbidden');
			assert.strictEqual(error.statusCode, 403);
			assert.strictEqual(error.message, 'Forbidden');
		});

		it('should set default message', () => {
			const error = new ForbiddenError();
			assert.strictEqual(error.statusCode, 403);
			assert.strictEqual(error.message, 'Forbidden');
		});
	});

	describe('UnauthorizedError', { concurrency: true }, () => {
		it('should set the statusCode to 401 and message correctly', () => {
			const error = new UnauthorizedError('Unauthorized');
			assert.strictEqual(error.statusCode, 401);
			assert.strictEqual(error.message, 'Unauthorized');
		});

		it('should set default message', () => {
			const error = new UnauthorizedError();
			assert.strictEqual(error.statusCode, 401);
			assert.strictEqual(error.message, 'Unauthorized');
		});
	});

	describe('NotFoundError', { concurrency: true }, () => {
		it('should set the statusCode to 404 and message correctly', () => {
			const error = new NotFoundError('Not Found');
			assert.strictEqual(error.statusCode, 404);
			assert.strictEqual(error.message, 'Not Found');
		});

		it('should set default message', () => {
			const error = new NotFoundError();
			assert.strictEqual(error.statusCode, 404);
			assert.strictEqual(error.message, 'Not Found');
		});
	});

	describe('ValidationError', { concurrency: true }, () => {
		it('should set the statusCode to 422 and message correctly', () => {
			const error = new ValidationError('Validation Error');
			assert.strictEqual(error.statusCode, 422);
			assert.strictEqual(error.message, 'Validation Error');
		});

		it('should set default message', () => {
			const error = new ValidationError();
			assert.strictEqual(error.statusCode, 422);
			assert.strictEqual(error.message, 'Validation Error');
		});
	});

	describe('UnimplementedFunctionError', { concurrency: true }, () => {
		it('should set the statusCode to 501 and message correctly', () => {
			const error = new UnimplementedFunctionError('Function Not Implemented');
			assert.strictEqual(error.statusCode, 501);
			assert.strictEqual(error.message, 'Function Not Implemented');
		});

		it('should set default message', () => {
			const error = new UnimplementedFunctionError();
			assert.strictEqual(error.statusCode, 501);
			assert.strictEqual(error.message, 'Function Not Implemented');
		});
	});
});

File: cmd/error_test.go
package main

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestErrorPages(t *testing.T) {
	tests := []struct {
		name       string
		statusCode int
		statusText string
		message    string
	}{
		{"bad request", http.StatusBadRequest, "Bad Request", "The request was invalid"},
		{"forbidden", http.StatusForbidden, "Forbidden", "Access is not allowed"},
		{"not found", http.StatusNotFound, "Not Found", "The page could not be found"},
		{"server error", http.StatusInternalServerError, "Internal Server Error", "Something went wrong"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			req := httptest.NewRequest(http.MethodGet, "http://commit.jaw.dev/missing", nil)
			req.Header.Set("User-Agent", "Mozilla/5.0")
			rr := httptest.NewRecorder()

			respond(rr, req, tt.statusCode, tt.message)

			if rr.Code != tt.statusCode {
				t.Fatalf("status = %d, want %d", rr.Code, tt.statusCode)
			}
			if got := rr.Header().Get("Content-Type"); got != "text/html; charset=utf-8" {
				t.Errorf("Content-Type = %q, want text/html; charset=utf-8", got)
			}

			body := rr.Body.String()
			for _, want := range []string{
				"<!DOCTYPE html>",
				tt.statusText,
				tt.message,
				"Back home",
				"Report an issue",
			} {
				if !strings.Contains(body, want) {
					t.Errorf("response does not contain %q", want)
				}
			}
		})
	}
}

func TestNotFoundRouteUsesErrorPage(t *testing.T) {
	app := newTestApp(&mockGenerator{})
	req := httptest.NewRequest(http.MethodGet, "http://commit.jaw.dev/missing", nil)
	req.Header.Set("User-Agent", "Mozilla/5.0")
	rr := httptest.NewRecorder()

	app.routes().ServeHTTP(rr, req)

	if rr.Code != http.StatusNotFound {
		t.Fatalf("status = %d, want %d", rr.Code, http.StatusNotFound)
	}
	if !strings.Contains(rr.Body.String(), "<h1>Not Found</h1>") {
		t.Error("response does not contain the styled not-found page")
	}
}

func TestErrorResponseJSON(t *testing.T) {
	tests := []struct {
		name      string
		setHeader func(*http.Request)
	}{
		{
			name: "accept header",
			setHeader: func(r *http.Request) {
				r.Header.Set("Accept", "application/json")
			},
		},
		{
			name: "curl user agent",
			setHeader: func(r *http.Request) {
				r.Header.Set("User-Agent", "curl/8.0.0")
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			req := httptest.NewRequest(http.MethodGet, "http://commit.jaw.dev/missing", nil)
			tt.setHeader(req)
			rr := httptest.NewRecorder()

			respond(rr, req, http.StatusNotFound, "The page could not be found")

			if rr.Code != http.StatusNotFound {
				t.Fatalf("status = %d, want %d", rr.Code, http.StatusNotFound)
			}
			if got := rr.Header().Get("Content-Type"); got != "application/json" {
				t.Errorf("Content-Type = %q, want application/json", got)
			}

			var body map[string]string
			if err := json.NewDecoder(rr.Body).Decode(&body); err != nil {
				t.Fatal(err)
			}
			if body["message"] != "The page could not be found" {
				t.Errorf("message = %q, want %q", body["message"], "The page could not be found")
			}
		})
	}
}

func TestErrorPageEscapesMessage(t *testing.T) {
	req := httptest.NewRequest(http.MethodGet, "http://commit.jaw.dev/missing", nil)
	rr := httptest.NewRecorder()

	respond(rr, req, http.StatusBadRequest, `<script>alert("no")</script>`)

	if strings.Contains(rr.Body.String(), `<script>alert("no")</script>`) {
		t.Error("response contains an unescaped error message")
	}
	if !strings.Contains(rr.Body.String(), "&lt;script&gt;") {
		t.Error("response does not contain the escaped error message")
	}
}
